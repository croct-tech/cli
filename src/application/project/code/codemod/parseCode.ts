import {Node} from '@babel/types';
import generate from '@babel/generator';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/code/codemod/codemod';
import {Language, parse} from '@/application/project/code/codemod/parser';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<Node, O>,
    languages: Language[],
};

export class ParseCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<Node, O>;

    private readonly languages: Language[];

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
        this.languages = configuration.languages;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(parse(input, this.languages), options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: generate(
                result.result,
                {
                    compact: false,
                    retainFunctionParens: true,
                },
            ).code,
        };
    }
}
