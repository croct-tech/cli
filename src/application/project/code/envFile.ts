import {FileSystem} from '@/application/fs/fileSystem';

export class EnvFile {
    private readonly fileSystem: FileSystem;

    private readonly path: string;

    public constructor(fileSystem: FileSystem, path: string) {
        this.fileSystem = fileSystem;
        this.path = path;
    }

    public getName(): string {
        return this.fileSystem.getBaseName(this.path);
    }

    public exists(): Promise<boolean> {
        return this.fileSystem.exists(this.path);
    }

    public async hasVariable(name: string): Promise<boolean> {
        const content = await this.read();

        if (content === '') {
            return false;
        }

        const escapedVariable = EnvFile.escapeRegex(name);
        const regex = new RegExp(`^${escapedVariable}\\s*=`, 'm');

        return regex.test(content);
    }

    public async setVariables(variables: Record<string, string>): Promise<void> {
        const content = await this.read();

        try {
            for (const [name, value] of Object.entries(variables)) {
                await this.setVariable(name, value);
            }
        } catch (error) {
            await this.write(content);

            throw error;
        }
    }

    public async setVariable(name: string, value: string): Promise<void> {
        const content = await this.read();
        const escapedName = EnvFile.escapeRegex(name);

        if (content === '') {
            return this.write(`${name}=${value}`);
        }

        const entry = `${name}=${value}`;

        const updatedContent = content.replace(
            new RegExp(
                `${escapedName}\\s*=\\s*((?!['"\`]).*$|\`(?:\\.|[^\`])*\`|'(?:\\.|[^'])*'|"(?:\\.|[^"])*")`,
                'm',
            ),
            entry,
        );

        if (updatedContent !== content || updatedContent.includes(entry)) {
            return this.write(updatedContent);
        }

        return this.write(`${content}${content.endsWith('\n') ? '' : '\n'}${name}=${value}`);
    }

    private async write(content: string): Promise<void> {
        await this.fileSystem.writeTextFile(this.path, content, {
            overwrite: true,
        });
    }

    private async read(): Promise<string> {
        if (!await this.exists()) {
            return '';
        }

        return this.fileSystem.readTextFile(this.path);
    }

    private static escapeRegex(value: string): string {
        return value.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
    }
}
