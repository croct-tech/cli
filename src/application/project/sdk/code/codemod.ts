export type ResultCode<T> = {
    modified: boolean,
    result: T,
};

export type CodemodOptions = Record<string, any>;

export interface Codemod<I, O extends CodemodOptions = CodemodOptions> {
    apply(input: I, options?: O): Promise<ResultCode<I>>;
}

export class CodemodError extends Error {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, CodemodError.prototype);
    }
}

export class MalformedCodeError extends CodemodError {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, MalformedCodeError.prototype);
    }
}
