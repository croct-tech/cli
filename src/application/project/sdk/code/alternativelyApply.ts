import {Codemod, CodemodOptions, ResultCode} from '@/application/project/sdk/code/codemod';

export class AlternativelyApply<T, O extends CodemodOptions> implements Codemod<T, O> {
    private readonly codemods: Array<Codemod<T, O>>;

    public constructor(codemod: Codemod<T, O>, ...codemods: Array<Codemod<T, O>>) {
        this.codemods = [codemod, ...codemods];
    }

    public async apply(input: T, options?: O): Promise<ResultCode<T>> {
        for (const codemod of this.codemods) {
            const result = await codemod.apply(input, options);

            if (result.modified) {
                return result;
            }
        }

        return {
            modified: false,
            result: input,
        };
    }
}
