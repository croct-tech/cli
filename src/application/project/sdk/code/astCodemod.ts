import type {namedTypes as Ast} from 'ast-types';
import {print} from 'recast';
import {Codemod, ResultCode} from '@/application/project/sdk/code/transformation';
import {parse} from '@/application/project/sdk/code/parser';

export class AstCodemod implements Codemod<string> {
    private readonly codemods: Array<Codemod<Ast.Node>>;

    public constructor(...transformers: Array<Codemod<Ast.Node>>) {
        this.codemods = transformers;
    }

    public apply(input: string): ResultCode<string> {
        if (this.codemods.length === 0) {
            return {
                modified: false,
                result: input,
            };
        }

        const ast = parse(input, ['jsx', 'typescript']);

        const result = this.codemods.reduce<ResultCode<Ast.Node>>(
            (next, transformer) => {
                const transformation = transformer.apply(next.result);

                return {
                    modified: transformation.modified || next.modified,
                    result: transformation.result,
                };
            },
            {
                modified: false,
                result: ast,
            },
        );

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result).code,
        };
    }
}
