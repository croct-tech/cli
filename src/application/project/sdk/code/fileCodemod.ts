import {readFile, writeFile} from 'fs/promises';
import {Codemod, CodemodError, ResultCode} from '@/application/project/sdk/code/transformation';
import {formatCause} from '@/application/error';

export class FileCodemod implements Codemod<string> {
    private readonly codemod: Codemod<string>;

    public constructor(codemod: Codemod<string>) {
        this.codemod = codemod;
    }

    public async apply(input: string): Promise<ResultCode<string>> {
        let source = '';

        try {
            source = await readFile(input, 'utf-8');
        } catch (error) {
            if (error.code !== 'ENOENT') {
                throw new CodemodError(`Failed to read file: ${formatCause(error)}`);
            }
        }

        const result = await this.codemod.apply(source);

        if (result.modified) {
            try {
                await writeFile(input, result.result, {encoding: 'utf-8', flag: 'w'});
            } catch (error) {
                throw new CodemodError(`Failed to write file: ${formatCause(error)}`);
            }
        }

        return result;
    }
}
