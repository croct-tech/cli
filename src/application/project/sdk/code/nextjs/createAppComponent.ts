import {namedTypes as Ast, builders as builder} from 'ast-types';
import {Codemod, ResultCode} from '@/application/project/sdk/code/codemod';
import {hasImport} from '@/application/project/sdk/code/javascript/hasImport';

export type AppComponentOptions = {
    typescript?: boolean,
};

export class CreateAppComponent implements Codemod<Ast.File, AppComponentOptions> {
    public apply(input: Ast.File, options: AppComponentOptions = {}): Promise<ResultCode<Ast.File>> {
        const isTypescript = options.typescript ?? false;
        const {body} = input.program;

        if (!CreateAppComponent.hasImport(input, '@croct/plug-next/CroctProvider', 'CroctProvider')) {
            body.unshift(CreateAppComponent.import('value', 'CroctProvider', '@croct/plug-next/CroctProvider'));
        }

        if (isTypescript) {
            if (!CreateAppComponent.hasImport(input, 'next/app', 'AppProps')) {
                body.unshift(CreateAppComponent.import('type', 'AppProps', 'next/app'));
            }

            if (!CreateAppComponent.hasImport(input, 'react', 'ReactNode')) {
                body.unshift(CreateAppComponent.import('type', 'ReactNode', 'react'));
            }
        }

        body.push(
            builder.exportDefaultDeclaration.from({
                declaration: builder.functionDeclaration.from({
                    id: builder.identifier('App'),
                    params: [
                        builder.objectPattern.from({
                            properties: [
                                builder.property.from({
                                    kind: 'init',
                                    key: builder.identifier('Component'),
                                    value: builder.identifier('Component'),
                                }),
                                builder.property.from({
                                    kind: 'init',
                                    key: builder.identifier('pageProps'),
                                    value: builder.identifier('pageProps'),
                                }),
                            ],
                            typeAnnotation: isTypescript
                                ? builder.tsTypeAnnotation.from({
                                    typeAnnotation: builder.tsTypeReference.from({
                                        typeName: builder.identifier('AppProps'),
                                    }),
                                })
                                : null,
                        }),
                    ],
                    returnType: isTypescript
                        ? builder.tsTypeAnnotation.from({
                            typeAnnotation: builder.tsTypeReference.from({
                                typeName: builder.identifier('ReactNode'),
                            }),
                        })
                        : null,
                    body: builder.blockStatement([
                        builder.returnStatement(
                            builder.jsxElement.from({
                                openingElement: builder.jsxOpeningElement.from({
                                    name: builder.jsxIdentifier('CroctProvider'),
                                    attributes: [],
                                }),
                                closingElement: builder.jsxClosingElement.from({
                                    name: builder.jsxIdentifier('CroctProvider'),
                                }),
                                children: [
                                    builder.jsxText('\n'),
                                    builder.jsxElement.from({
                                        openingElement: builder.jsxOpeningElement.from({
                                            name: builder.jsxIdentifier('Component'),
                                            attributes: [
                                                builder.jsxSpreadAttribute.from({
                                                    argument: builder.identifier('pageProps'),
                                                }),
                                            ],
                                        }),
                                        closingElement: builder.jsxClosingElement.from({
                                            name: builder.jsxIdentifier('Component'),
                                        }),
                                        children: [],
                                    }),
                                    builder.jsxText('\n'),
                                ],
                            }),
                        ),
                    ]),
                }),
            }),
        );

        return Promise.resolve({
            modified: true,
            result: input,
        });
    }

    private static hasImport(input: Ast.File, moduleName: string, importName: string): boolean {
        return hasImport(input, {
            moduleName: moduleName,
            importName: importName,
            localName: importName,
        });
    }

    private static import(type: 'value' | 'type', name: string, from: string): Ast.ImportDeclaration {
        return builder.importDeclaration.from({
            importKind: type,
            specifiers: [
                builder.importSpecifier.from({
                    imported: builder.identifier(name),
                }),
            ],
            source: builder.literal(from),
        });
    }
}
