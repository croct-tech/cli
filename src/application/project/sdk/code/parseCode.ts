import type {namedTypes as Ast} from 'ast-types';
import {print} from 'recast';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/sdk/code/codemod';
import {Language, parse} from '@/application/project/sdk/code/parser';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<Ast.Node, O>,
    languages: Language[],
};

export class ParseCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<Ast.Node, O>;

    private readonly languages: Language[];

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
        this.languages = configuration.languages;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(parse(input, this.languages), options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result).code,
        };
    }
}
