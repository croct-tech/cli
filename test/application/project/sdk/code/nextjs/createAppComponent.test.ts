import {resolve} from 'path';
import {loadFixtures} from '../fixtures';
import {ParseCode} from '@/application/project/sdk/code/parseCode';
import {AppComponentOptions, CreateAppComponent} from '@/application/project/sdk/code/nextjs/createAppComponent';

describe('CreateAppComponent', () => {
    const scenarios = loadFixtures<AppComponentOptions>(
        resolve(__dirname, '../fixtures/nextjs-app-component'),
        {},
        {},
    );

    it.each(scenarios)('should correctly transform $name', async ({name, fixture}) => {
        const transformer = new ParseCode({
            languages: ['typescript', 'jsx'],
            codemod: new CreateAppComponent({
                provider: {
                    component: 'CroctProvider',
                    module: '@croct/plug-next/CroctProvider',
                },
            }),
        });

        const output = await transformer.apply(fixture, {
            typescript: name.endsWith('.tsx'),
        });

        expect(output.result).toMatchSnapshot(name);
    });
});
