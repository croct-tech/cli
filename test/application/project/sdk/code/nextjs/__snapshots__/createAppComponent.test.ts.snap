// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAppComponent should correctly transform allImports.tsx: allImports.tsx 1`] = `
"import type { ReactNode } from "react";
import type { AppProps } from "next/app";
import { CroctProvider } from "@croct/plug-next/CroctProvider";

export default function App(
  {
    Component: Component,
    pageProps: pageProps
  }: AppProps
): ReactNode {
  return (
    <CroctProvider>
      <Component {...pageProps}></Component>
    </CroctProvider>
  );
}
"
`;

exports[`CreateAppComponent should correctly transform empty.jsx: empty.jsx 1`] = `
"import { CroctProvider } from "@croct/plug-next/CroctProvider";

export default function App(
  {
    Component: Component,
    pageProps: pageProps
  }
) {
  return (
    <CroctProvider>
      <Component {...pageProps}></Component>
    </CroctProvider>
  );
}"
`;

exports[`CreateAppComponent should correctly transform empty.tsx: empty.tsx 1`] = `
"import type { ReactNode } from "react";
import type { AppProps } from "next/app";
import { CroctProvider } from "@croct/plug-next/CroctProvider";

export default function App(
  {
    Component: Component,
    pageProps: pageProps
  }: AppProps
): ReactNode {
  return (
    <CroctProvider>
      <Component {...pageProps}></Component>
    </CroctProvider>
  );
}"
`;

exports[`CreateAppComponent should correctly transform providerImport.jsx: providerImport.jsx 1`] = `
"import { CroctProvider } from "@croct/plug-next/CroctProvider";

export default function App(
  {
    Component: Component,
    pageProps: pageProps
  }
) {
  return (
    <CroctProvider>
      <Component {...pageProps}></Component>
    </CroctProvider>
  );
}
"
`;

exports[`CreateAppComponent should correctly transform unrelatedCode.tsx: unrelatedCode.tsx 1`] = `
"import type { ReactNode } from "react";
import type { AppProps } from "next/app";
import { CroctProvider } from "@croct/plug-next/CroctProvider";
import {Fragment} from "react";

export function UnrelatedComponent() {
    return <Fragment />;
}

export default function App(
    {
        Component: Component,
        pageProps: pageProps
    }: AppProps
): ReactNode {
    return (
        <CroctProvider>
            <Component {...pageProps}></Component>
        </CroctProvider>
    );
}
"
`;
